src/analyzer.ts

```diff
@@ -0,0 +1,766 @@
+import {Descriptor, ElementDescriptor, FeatureDescriptor, BehaviorDescriptor} from './ast-utils/descriptors';
+
+function reduceMetadata(m1:DocumentDescriptor, m2:DocumentDescriptor): DocumentDescriptor {
+  return {
+    elements:  m1.elements.concat(m2.elements),
+    features:  m1.features.concat(m2.features),
+    behaviors: m1.behaviors.concat(m2.behaviors),
+  };
+}
+
+var EMPTY_METADATA: DocumentDescriptor = {elements: [], features: [], behaviors: []};
+
+/**
+ * Package of a parsed JS script
+ */
+interface ParsedJS {
+  ast: estree.Program;
+  scriptElement: dom5.Node;
+}
+
+/**
+ * The metadata for all features and elements defined in one document
+ */
+interface DocumentDescriptor {
+  /**
+   * The elements from the document.
+   */
+  elements: ElementDescriptor[];
+
+  /**
+   * The features from the document
+   */
+  features: FeatureDescriptor[];
+
+  /**
+   * The behaviors from the document
+   */
+  behaviors: BehaviorDescriptor[];
+
+  href?: string;
+
+  imports?: DocumentDescriptor[];
+
+  parsedScript?: estree.Program;
+
+  html?: ParsedImport;
+}
+
+/**
+ * The metadata of an entire HTML document, in promises.
+ */
+interface AnalyzedDocument {
+  /**
+   * The url of the document.
+   */
+  href: string;
+  /**
+   * The parsed representation of the doc. Use the `ast` property to get
+   * the full `parse5` ast.
+   */
+  htmlLoaded: Promise<ParsedImport>;
+
+  /**
+   * Resolves to the list of this Document's transitive import dependencies.
+   */
+  depsLoaded: Promise<string[]>;
+
+  /**
+   * The direct dependencies of the document.
+   */
+  depHrefs: string[];
+  /**
+   * Resolves to the list of this Document's import dependencies
+   */
+  metadataLoaded: Promise<DocumentDescriptor>;
+}
+
+/**
+ * Options for `Analyzer.analzye`
+ */
+interface LoadOptions {
+  /**
+   * Whether `annotate()` should be skipped.
+   */
+  noAnnotations?: boolean;
+  /**
+   * Content to resolve `href` to instead of loading from the file system.
+   */
+  content?: string;
+  /**
+   * Whether the generated descriptors should be cleaned of redundant data.
+   */
+  clean?: boolean;
+  /**
+   * `xhr` to use XMLHttpRequest.
+   * `fs` to use the local filesystem.
+   * `permissive` to use the local filesystem and return empty files when a
+   * path can't be found.
+   * Default is `fs` in node and `xhr` in the browser.
+   */
+  resolver?: string;
+  /**
+   * A predicate function that indicates which files should be ignored by
+   * the loader. By default all files not located under the dirname
+   * of `href` will be ignored.
+   */
+  filter?: (path:string)=> boolean;
+}
+
+/**
+ * An Error extended with location metadata.
+ */
+interface LocError extends Error{
+  location: {line: number; column: number};
+  ownerDocument: string;
+}
+
+/**
+ * A database of Polymer metadata defined in HTML
+ *
+ * @param  {boolean} attachAST  If true, attach a parse5 compliant AST
```

Comment:
Also, is there any hope of getting tsc to generate the jsdoc/jscompiler type annotations?